### PR TITLE
Correctly detect the target directory

### PR DESCRIPTION
**Pull request by DJMcNab**
The only way cargo provides to get the target directory, is to use `cargo metadata`.
Requiring that we get a directory at the project root adds additional clutter. 
For reference, I'm want this for [vello](https://github.com/linebender/vello), which uses `run-wasm` for examples, and therefore puts it in the `examples` directory.

This should also ensure that we obey the wishes of any users who wish to use a global target directory.

### CODE REVIEW

**Comment by DJMcNab** [comment]
I have now tested this in vello on windows, and it works as expected.

**Comment by rukai** [comment]
Thankyou for raising this, generating an extra target directory is a genuine issue.

Unfortunately `cargo metadata` is really expensive! I measured 300ms against vello on the laptop im currently using.
I wonder if we could use the current method and then fallback to a more robust method if the target directory is missing?
Maybe it would be enough to traverse up the directory tree until we find the outermost Cargo.toml.
Or maybe we really do need to call `cargo metadata`

Additionally if we do use the `cargo metadata` solution we should process a `serde_json::Value` instead of introducing the `serde_derive` proc macro as that will affect overall build times as the proc macro needs to finish compiling before cargo-run-wasm can start compiling.

Sorry to be annoying about this stuff but cargo-run-wasm being fast to run and even faster to rerun is an important goal of the project.
I'm happy to take over the investigation if this ends up being more than you bargained for when you opened the PR.

**Comment by DJMcNab** [comment]
To clarify, when you checked `cargo metadata`, I'm presuming that was with `--no-deps` set?

It is very frustrating that cargo doesn't provide a way to just get this value directly.

The shortcut of checking one level up the directory tree is a good idea, I'll implement that.

**Comment by rukai** [comment]
Oh I'll have to try --no-deps, sorry I missed that

**Comment by rukai** [comment]
I can confirm --no-deps is significantly better: `~80ms`
Its still not ideal though, so a shortcut would be appreciated.

**Comment by DJMcNab** [comment]
I've now implemented the shortcut, which should execute concurrently with `cargo metadata` - the shortcut has priority to ensure that we get a consistent result, even though it might sometimes be wrong.

File: src/lib.rs [hunk 1]
```diff
@@ -161,25 +163,19 @@ pub fn run_wasm_with_css(css: &str) {
...
-        // This results in a separate rustflags value for native and wasm builds.
-        // Currently rust triggers a full rebuild every time the rustflags value changes.
-        //
-        // Therefore we have this hack where we use a different target dir for wasm builds to avoid constantly triggering full rebuilds.
-        // When this issue is resolved we might be able to remove this hack: https://github.com/rust-lang/cargo/issues/8716
-        "--target-dir",
-        "target/wasm-examples-target",
-    ];
+
+    let project_root = Path::new(&std::env::var("CARGO_MANIFEST_DIR").unwrap()).to_path_buf();
```
Comment (rukai) [inline src/lib.rs:169]: project_root is used as the working directory for cargo build. Changing the working directory like this breaks cargo-run-wasm as it is used in other projects such as wgpu. I would recommend verifying that wgpu or winits use of cargo-run-wasm continues to function before pushing again.

File: src/lib.rs [hunk 2]
```diff
@@ -161,25 +163,19 @@ pub fn run_wasm_with_css(css: &str) {
 
     // build wasm example via cargo
     let cargo = env::var("CARGO").unwrap_or_else(|_| "cargo".to_string());
-    let project_root = Path::new(&std::env::var("CARGO_MANIFEST_DIR").unwrap())
-        .ancestors()
-        .nth(1)
-        .unwrap()
-        .to_path_buf();
-    let mut cargo_args = vec![
```
Comment (rukai) [inline src/lib.rs:169]: I would prefer to keep the vec constructed in a single go like this rather than using `.extend`

File: src/target_dir.rs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+            .get("target_directory")
+            .expect("Should have target directory")
+            .as_str()
+            .unwrap()
+            .to_owned();
+        Self { target_directory }
+    }
+}
+
+pub fn get_target_directory(cargo_executable: &str, manifest_dir: &Path) -> String {
```
Comment (rukai) [inline src/target_dir.rs]: this should return a PathBuf instead of a string

**Review by rukai — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 1 above)
Comment (DJMcNab) [reply src/lib.rs:169]: This surprises me. My mental model of cargo's building model within workspaces is that it doesn't matter where in the workspace the command is run, it always has the same effect.

In my mind, choosing an arbitrary directory as working directory is strange. The only information we have about which workspace we are in is `CARGO_MANIFEST_DIR`, so it seemed that we should use this.

I'll see where my mental model breaks down by trying this in winit and wgpu.

**Comment by DJMcNab** [comment]
I've now tested with wgpu, winit and vello; all three of which now work correctly.

**Review by rukai — APPROVED** [review]
Thankyou, the regression is fixed so I'll go ahead and merge.
There are a few minor cleanups I would like but they are not important so I'll do them in a follow up PR.

**Pull request merged**